rust/capture/src/v0_endpoint.rs

```diff
@@ -226,3 +326,53 @@ pub async fn process_events<'a>(
         sink.send_batch(events).await
     }
 }
+
+#[instrument(skip_all, fields(events = events.len()))]
+pub async fn process_replay_events<'a>(
+    sink: Arc<dyn sinks::Event + Send + Sync>,
+    events: &'a [RawEvent],
+    context: &'a ProcessingContext,
+) -> Result<(), CaptureError> {
+    let snapshot_items: Vec<Value> = events
+        .iter()
+        .map(|e| match e.properties.get("$snapshot_data") {
+            Some(Value::Array(value)) => Ok(value.to_vec()),
+            Some(Value::Object(value)) => Ok([Value::Object(value.clone())].to_vec()),
+            _ => Err(CaptureError::MissingSnapshotData),
+        })
+        .collect::<Result<Vec<Vec<_>>, CaptureError>>()?
+        .into_iter()
+        .flatten()
+        .collect();
+
+    let session_id = events[0]
+        .properties
+        .get("$session_id")
+        .ok_or(CaptureError::MissingSessionId)?;
+    let window_id = events[0]
+        .properties
+        .get("$window_id")
+        .ok_or(CaptureError::MissingWindowId)?;
+    let event = ProcessedEvent {
+        data_type: DataType::SnapshotMain,
+        uuid: events[0].uuid.unwrap_or_else(uuid_v7),
+        distinct_id: events[0].extract_distinct_id()?,
+        ip: context.client_ip.clone(),
+        data: json!({
+            "event": "$snapshot_items",
+            "properties": {
+                "distinct_id": events[0].extract_distinct_id()?,
+                "$session_id": session_id,
+                "$window_id": window_id,
+                "$snapshot_source": events[0].properties.get("$snapshot_source").unwrap_or(&Value::String(String::from("web"))),
+                "$snapshot_items": snapshot_items,
+            }
+        }).to_string(),
+        now: context.now.clone(),
+        sent_at: context.sent_at,
+        token: context.token.clone(),
+        session_id: Some(session_id.as_str().unwrap().to_string()),
```

Comment:
I'd really rather not `unwrap()` here. Could we change line 348 onwards to something like:
```rust
    let session_id = events[0]
        .properties
        .get("$session_id")
        .ok_or(CaptureError::MissingSessionId)?
        .as_str()
        .ok_or(CaptureError::InvalidSessionId)? // Adding this error type ofc
        .to_string();
```

Reply:
good catch, fixed